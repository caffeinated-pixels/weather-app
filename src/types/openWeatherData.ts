export type Coordinates = {
  latitude: number
  longitude: number
}

export type GeocodingApiResult = {
  name: string
  country: string
  lat: number
  lon: number
  state: string
  local_names: Record<string, string>
}

export type CurrentWeatherSummary = {
  id: number
  main: string
  description: string
  icon: string
}

type Temp = {
  min: number
  max: number
  day: number
  night: number
  eve: number
  morn: number
}

type FeelsLikeDaytimes = Omit<Temp, 'min' | 'max'>

type UnixTimeStamp = number

type SunTimes = {
  sunrise: UnixTimeStamp
  sunset: UnixTimeStamp
}

type MoonTimes = {
  moonrise: UnixTimeStamp
  moonset: UnixTimeStamp
  moon_phase: UnixTimeStamp
}

type SunAndMoonTimes = SunTimes & MoonTimes

type WindData = {
  wind_speed: number
  wind_deg: number
  wind_gust: number
}

type GeneralWeatherData = {
  dt: UnixTimeStamp
  temp: Temp
  feels_like: FeelsLikeDaytimes
  pressure: number
  humidity: number
  dew_point: number
  weather: CurrentWeatherSummary[]
  clouds: number
  pop: number
  uvi: number
}

export type DailyData = SunAndMoonTimes & WindData & GeneralWeatherData

export type HourlyData = WindData &
  GeneralWeatherData & {
    temp: number
    visibility: number
  }

type Minutely = {
  dt: UnixTimeStamp
  precipitation: number
}

type Daily = SunAndMoonTimes & WindData & GeneralWeatherData

type CurrentWeatherData = SunTimes &
  Omit<WindData, 'wind_gust'> &
  GeneralWeatherData & {
    temp: number
    visibility: number
    feels_like: number
  }

export type FullWeatherData = {
  lat: number
  lon: number
  timezone: string
  timezone_offset: number
  current: CurrentWeatherData
  minutely: Minutely[]
  hourly: HourlyData[]
  daily: Daily[]
}

export type ProcessedWeatherData = {
  iconUrl: string
  iconAltText: string
  weatherDescription: string
  currentTemp: string
  minTemp: string
  maxTemp: string
  feelsLikeTemp: string
  windSpeed: string
  windDirection: string
  windIconRotation: number
  visibility: string
  precipProb: string
  humidity: string
  dewPoint: string
  uvIndex: string
  pressure: string
  hourlyBoxes: JSX.Element
  dailyBoxes: JSX.Element
}

export type WeatherDataResults = {
  weatherData: FullWeatherData | null
  processedWeatherData: ProcessedWeatherData | null
  isLoading: boolean
  isError: boolean
  errorMsg: string
}

export type LocationResults = {
  resultsArr: GeocodingApiResult[]
  searchComplete: boolean
  searchMatchFail: boolean
  apiError: boolean
  errorMsg: string
}

// location data that gets returned from the FullWeatherData query
export type LocationData = Coordinates & {
  city: string
  country: string
  state?: string
}
