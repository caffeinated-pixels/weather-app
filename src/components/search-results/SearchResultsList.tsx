import { HandleResultsChoice } from '../../WeatherDataContext'
import { GeocodingApiResult } from '../../types/openWeatherData'

type SearchResultsProps = {
  resultsArr: GeocodingApiResult[]
  handleResultsChoice: HandleResultsChoice
}

export const SearchResultsList = ({
  resultsArr,
  handleResultsChoice,
}: SearchResultsProps) => {
  return (
    <ul className="search-result-list">
      {resultsArr.map((location: GeocodingApiResult, i: number) => {
        const backgroundColor = i % 2 === 0 ? 'stripe-light' : 'stripe-dark'

        return (
          <li
            key={`result${i}`}
            className={`search-result ${backgroundColor}`}
            onClick={() => handleResultsChoice(i)}
            onKeyDown={(e) => handleResultsChoice(i, e)}
            tabIndex={0}
          >
            {location.name}, {location.state && `${location.state}, `}
            {location.country}
          </li>
        )
      })}
    </ul>
  )
}
