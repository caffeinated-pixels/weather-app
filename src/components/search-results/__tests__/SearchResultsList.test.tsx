import '@testing-library/jest-dom'

import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import renderer from 'react-test-renderer'

import { GeocodingApiResult } from '../../../types/openWeatherData'
import { SearchResultsList } from '../SearchResultsList'

export const mockResultsArr: GeocodingApiResult[] = [
  {
    name: 'Brighton',
    local_names: {
      yi: 'ברייטאן',
      sw: 'Brighton',
    },
    lat: 50.8214626,
    lon: -0.1400561,
    country: 'GB',
    state: 'England',
  },
  {
    name: 'Brighton',
    local_names: { en: 'Brighton' },
    lat: 39.983721,
    lon: -104.8110775,
    country: 'US',
    state: 'Colorado',
  },
  {
    name: 'Brighton',
    lat: 49.542751,
    lon: -55.6406218,
    country: 'CA',
    state: 'Newfoundland and Labrador',
  },
  {
    name: 'Brighton',
    lat: 33.4342755,
    lon: -86.9472148,
    country: 'US',
    state: 'Alabama',
  },
  {
    name: 'Brighton',
    lat: 39.039773,
    lon: -90.1410023,
    country: 'US',
    state: 'Illinois',
  },
]

const mockHandleResultsChoice = jest.fn()
const renderSearchResultsList = () => {
  return render(
    <SearchResultsList
      resultsArr={mockResultsArr}
      handleResultsChoice={mockHandleResultsChoice}
    />
  )
}

describe('SearchResultsList', () => {
  it('should match the snapshot', () => {
    const domTree = renderer
      .create(
        <SearchResultsList
          resultsArr={mockResultsArr}
          handleResultsChoice={mockHandleResultsChoice}
        />
      )
      .toJSON()
    expect(domTree).toMatchSnapshot()
  })

  it('should render the SearchResultsList component', () => {
    renderSearchResultsList()

    expect(screen.getByRole('list')).toBeInTheDocument()
  })

  it('should render the correct number of results', () => {
    renderSearchResultsList()

    expect(screen.getAllByRole('listitem').length).toBe(5)
  })

  it('should render the correct text for each result', () => {
    renderSearchResultsList()

    const listItems = screen.getAllByRole('listitem')

    listItems.forEach((item, i) => {
      expect(item).toHaveTextContent(
        `${mockResultsArr[i].name}, ${
          mockResultsArr[i].state && `${mockResultsArr[i].state}, `
        }${mockResultsArr[i].country}`
      )
    })
  })

  it('should call handleResultsChoice when a result is clicked', async () => {
    renderSearchResultsList()

    const listItems = screen.getAllByRole('listitem')

    await userEvent.click(listItems[0])
    expect(mockHandleResultsChoice).toHaveBeenCalled()

    for (let i = 0; i < listItems.length; i++) {
      await userEvent.click(listItems[i])
      expect(mockHandleResultsChoice).toHaveBeenCalledWith(i)
    }
  })

  it('should call handleResultsChoice when a result is focused and enter is pressed', async () => {
    renderSearchResultsList()

    const listItems = screen.getAllByRole('listitem')

    for (let i = 0; i < listItems.length; i++) {
      listItems[i].focus()
      await userEvent.keyboard('{enter}')
      expect(mockHandleResultsChoice).toHaveBeenCalled()
    }
  })
})
