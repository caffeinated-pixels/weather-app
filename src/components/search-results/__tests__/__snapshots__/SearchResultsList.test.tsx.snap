// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResultsList should match the snapshot 1`] = `
<ul
  className="search-result-list"
>
  <li
    className="search-result stripe-light"
    onClick={[Function]}
    onKeyDown={[Function]}
    tabIndex={0}
  >
    Brighton
    , 
    England, 
    GB
  </li>
  <li
    className="search-result stripe-dark"
    onClick={[Function]}
    onKeyDown={[Function]}
    tabIndex={0}
  >
    Brighton
    , 
    Colorado, 
    US
  </li>
  <li
    className="search-result stripe-light"
    onClick={[Function]}
    onKeyDown={[Function]}
    tabIndex={0}
  >
    Brighton
    , 
    Newfoundland and Labrador, 
    CA
  </li>
  <li
    className="search-result stripe-dark"
    onClick={[Function]}
    onKeyDown={[Function]}
    tabIndex={0}
  >
    Brighton
    , 
    Alabama, 
    US
  </li>
  <li
    className="search-result stripe-light"
    onClick={[Function]}
    onKeyDown={[Function]}
    tabIndex={0}
  >
    Brighton
    , 
    Illinois, 
    US
  </li>
</ul>
`;
